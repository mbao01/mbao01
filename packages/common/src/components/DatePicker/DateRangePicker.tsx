"use client";

import { useState } from "react";
import { format } from "date-fns";
import { type DateRange } from "react-day-picker";
import { CalendarIcon } from "@radix-ui/react-icons";
import { type DateRangePickerProps } from "./types";
import { cn } from "../../utilities";
import { Button } from "../Button";
import { Calendar } from "../Calendar";
import { Popover } from "../Popover";

export const DateRangePicker = ({
  name,
  size,
  wide,
  label,
  outline,
  variant,
  disabled,
  defaultRange,
  getRangeValue = (range) => ({
    from: range?.from?.toUTCString(),
    to: range?.to?.toUTCString(),
  }),
  getRangeLabel = (range) =>
    range?.from ? (
      range.to ? (
        <>
          {format(range.from, "LLL dd, y")} - {format(range.to, "LLL dd, y")}
        </>
      ) : (
        format(range.from, "LLL dd, y")
      )
    ) : undefined,
  ...props
}: DateRangePickerProps) => {
  const [range, setRange] = useState<DateRange | undefined>(defaultRange);

  const rangeLabel = getRangeLabel(range);
  const rangeValue = getRangeValue(range);

  return (
    <Popover>
      {name && (
        <>
          <input
            hidden
            className="hidden"
            aria-hidden="true"
            name={`${name}.from`}
            defaultValue={rangeValue.from}
          />
          <input
            hidden
            className="hidden"
            aria-hidden="true"
            name={`${name}.to`}
            defaultValue={rangeValue.to}
          />
        </>
      )}
      <Popover.Trigger asChild>
        <Button
          name={name}
          size={size}
          wide={wide}
          outline={outline}
          variant={variant}
          disabled={disabled}
          className={cn("justify-start flex-nowrap", !range && "font-normal")}
        >
          <span className="text-left line-clamp-1">
            {rangeLabel ?? label ?? "Pick a range"}
          </span>
          <CalendarIcon className="ml-auto h-4 w-4 opacity-50" />
        </Button>
      </Popover.Trigger>
      <Popover.Content className="w-auto p-0">
        <Calendar
          initialFocus
          mode="range"
          selected={range}
          onSelect={setRange}
          captionLayout="dropdown-buttons"
          {...props}
        />
      </Popover.Content>
    </Popover>
  );
};
