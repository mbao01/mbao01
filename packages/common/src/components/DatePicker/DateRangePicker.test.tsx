import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DateRangePicker } from "./DateRangePicker";

describe("DateRangePicker", () => {
  it("shows the current month", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<DateRangePicker />);

    const button = screen.getByRole("button", { name: "Pick a range" });

    await user.click(button);

    const date = new Date();
    const month = date.toLocaleString("default", { month: "long" });
    const year = date.getFullYear();

    expect(screen.getByText(`${month} ${year}`)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows a default/preselected month and range", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <DateRangePicker
        defaultMonth={new Date("2022-03-01")}
        defaultRange={{
          from: new Date("2022-03-01"),
          to: new Date("2022-03-15"),
        }}
      />
    );

    const button = screen.getByRole("button", {
      name: "Mar 01, 2022 - Mar 15, 2022",
    });

    await user.click(button);

    expect(screen.getByText("March 2022")).toBeInTheDocument();
    expect(screen.getByRole("gridcell", { name: "15" })).toHaveAttribute(
      "aria-selected",
      "true"
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows calendar with specific month", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <DateRangePicker defaultMonth={new Date("2022-03-01")} />
    );

    const button = screen.getByRole("button", { name: "Pick a range" });

    await user.click(button);

    expect(screen.getByText("March 2022")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a custom label", async () => {
    const { asFragment } = render(
      <DateRangePicker label="How long will you be staying?" />
    );

    expect(
      screen.getByRole("button", { name: "How long will you be staying?" })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("is disabled", async () => {
    const { asFragment } = render(<DateRangePicker disabled />);

    expect(screen.getByRole("button", { name: "Pick a range" })).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });
});
