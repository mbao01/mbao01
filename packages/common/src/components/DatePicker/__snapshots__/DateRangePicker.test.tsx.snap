// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateRangePicker > has a custom label 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r3»"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="closed"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      How long will you be staying?
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`DateRangePicker > is disabled 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r4»"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="closed"
    disabled=""
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick a range
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`DateRangePicker > shows a default/preselected month and range 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r1»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Mar 01, 2022 - Mar 15, 2022
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`DateRangePicker > shows calendar with specific month 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r2»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick a range
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`DateRangePicker > shows the current month 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r0»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick a range
    </span>
    <svg
      aria-hidden="true"
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;
